Lazy Import for Diffusers

DN6 opened the pull request:
# What does this PR do?
Adds lazy import functionality to Diffusers, similar to what exists in [Transformers](https://github.com/huggingface/transformers/blob/main/src/transformers/__init__.py)

### Benchmark:

Initially tested import speed up by running `time python -c "import diffusers"` with all backends (torch, transformers, flax, onnxruntime) installed. 

#### With Lazy Import
```
real    0m3.563s
user    0m3.420s
sys     0m0.963s
``` 

#### Without Lazy Import
```
real    0m5.391s
user    0m5.299s
sys     0m1.273s
```

This PR:

1. Adds lazy import to the following modules

```
diffusers
models
pipelines
schedulers
```

2. Moves objects defined in the `__init__.py` files of a module to their own dedicated files. e.g. `StableDiffusionPipelineOutput` is now in a `pipeline_output.py` file within the `stable_diffusion` module. This follows the same format as `transformers` and also keeps `__init__.py` reserved for import related code. 

TODO:
1. `utils/check_dummies` behaviour and the corresponding test is broken because of the new import structure. This is the last thing that needs to be addressed before merging.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)
#4260

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Core library:

- Schedulers: @williamberman and @patrickvonplaten
- Pipelines:  @patrickvonplaten and @sayakpaul
- Training examples: @sayakpaul and @patrickvonplaten
- Docs: @stevenliu and @yiyixu
- JAX and MPS: @pcuenca
- Audio: @sanchit-gandhi
- General functionalities: @patrickvonplaten and @sayakpaul

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- transformers: [different repo](https://github.com/huggingface/transformers)
- safetensors: [different repo](https://github.com/huggingface/safetensors)

-->

patrickvonplaten commented:
I'd actually do one big PR here and we can try to merge it quickly to not run into merge conflicts :-) 

We should measure the speed-up of importing things once this is merged and also try to clean up many of these imports inside functions in the `loaders.py` file, e.g.: https://github.com/huggingface/diffusers/blob/a76f2ad538e73b34d5fe7be08c8eb8ab38c7e90c/src/diffusers/loaders.py#L274

DN6 commented:
@patrickvonplaten  this is ready for another review. I ran into one blocker when testing locally. All the Fasts tests are passing with lazy imports, except for `tests/others/test_check_dummies.py`. 

`utils/check_dummies.py` is also affected by these changes. When parsing the init, the lack of real objects seems to break it. I'm a little stuck on the best way to fix it, if you have any suggestions, I'd appreciate it 👍🏽

DN6 commented:
@patrickvonplaten took a look into the import time profile. A big slowdown is happening because we import torch in our `utils` module if it is available. 
![lazy-import-profile](https://github.com/huggingface/diffusers/assets/7529846/bc620931-825e-477c-9b6d-a672795bfb8c)

`torch` is being imported from the `torch_utils` module
https://github.com/huggingface/diffusers/blob/705c592ea98ba4e288d837b9cba2767623c78603/src/diffusers/utils/__init__.py#L84

https://github.com/huggingface/diffusers/blob/705c592ea98ba4e288d837b9cba2767623c78603/src/diffusers/utils/torch_utils.py#L24

Changing this to avoid the import would give us a big boost. I can take a look into it.

patrickvonplaten commented:
Thanks for profiling this - very nice to get some insights here. Approach 1. looks good to me! 

I think by using a search & replace regex you can update all the files relatively quickly. E.g. you could update:

```py
from ...utils import randn_tensor
```

with the correct import statement in a regex and statements like:

```py
from ...utils import deprecate, logging, randn_tensor
```

can be updated to:

```py
from ...utils import deprecate, logging; from diffusers.utils.torch_utils import randn_tensor
```
and then `make style` will do the rest for you

patrickvonplaten commented on src/diffusers/pipelines/kandinsky2_2/__init__.py:
```diff
@@ -1,25 +1,50 @@
 
+_import_structure = {}
+_dummy_objects = {}
+
+_dummy_objects = {}
```
```suggestion
```

patrickvonplaten commented:
Think we can merge this very soon. Happy to help with the final failing tests if you'd like

DN6 commented:
@patrickvonplaten Yes if you could that would be awesome. Failing tests are related to `utils/check_copies` and `utils/check_dummies` and there is one SchedulerBase test related one. 

Here are the most recent benchmarks FYI. We're getting ~10X speed up as anticipated. 

Without Lazy Import (all backends installed)
```
real    0m4.513s
user    0m4.550s
sys     0m1.083s
```

With Lazy Import (all backends installed)
```
real    0m0.419s
user    0m0.706s
sys     0m0.531s
```

patrickvonplaten commented:
> @patrickvonplaten Yes if you could that would be awesome. Failing tests are related to `utils/check_copies` and `utils/check_dummies` and there is one SchedulerBase test related one.
> 
> Here are the most recent benchmarks FYI. We're getting ~10X speed up as anticipated.
> 
> Without Lazy Import (all backends installed)
> 
> ```
> real    0m4.513s
> user    0m4.550s
> sys     0m1.083s
> ```
> 
> With Lazy Import (all backends installed)
> 
> ```
> real    0m0.419s
> user    0m0.706s
> sys     0m0.531s
> ```

Nice! 

I'll try to get this one in today!

patrickvonplaten commented:
@DN6, I'm only seeing a 2x speed-up on my machine (maybe that's also not very representative though). I have `accelerate` installed as well (which is more or less required by now as well).

> python3 -c "import diffusers"  2.14s user 1.62s system 182% cpu 2.065 total
getting 2.14s for this PR

and getting:

>  python3 -c "import diffusers"  4.02s user 3.09s system 191% cpu 3.716 total

for "main"

(so more of a 2x speed-up)

Transformers is much faster for me. Any idea what could be going on here?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_4829). All of your documentation changes will be reflected on that endpoint.

DN6 commented:
> @DN6, I'm only seeing a 2x speed-up on my machine (maybe that's also not very representative though). I have `accelerate` installed as well (which is more or less required by now as well).
> 
> > python3 -c "import diffusers"  2.14s user 1.62s system 182% cpu 2.065 total
> > getting 2.14s for this PR
> 
> and getting:
> 
> > python3 -c "import diffusers"  4.02s user 3.09s system 191% cpu 3.716 total
> 
> for "main"
> 
> (so more of a 2x speed-up)
> 
> Transformers is much faster for me. Any idea what could be going on here?
 
 Taking a look into this.

DN6 commented:
@patrickvonplaten  Profiled imports with `accelerate` installed. Looks like if accelerate is available we import it, which imports torch and leads to the slow down.
![acclerate-lazy-import](https://github.com/huggingface/diffusers/assets/7529846/015a9091-61c4-4065-87c1-8decaba95bd1)
 
The `apply_forward_hook` from accelerate is only being used in the model files. So we can change the import there to `from utils.acclerate_utils import apply_forward_hook`  and remove the accelerate import from the utils init file.

DN6 commented on utils/check_dummies.py:
```diff
@@ -71,6 +71,9 @@ def read_init():
 
     # Get to the point we do the actual imports for type checking
     line_index = 0
+    while not lines[line_index].startswith("if TYPE_CHECKING"):
```
I can add this is in. For my own knowledge. What is TYPE_CHECKING doing?

patrickvonplaten replied:
It's essentially helping automatic type checkers to be able to understand all the imports: https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

DN6 commented:
@patrickvonplaten 
I added type checking to the diffusers init and moved the accelerate import out of the utils init and changed the relevant files. 

The import speed should be improved. Results on my machine with all backends + accelerate installed. 
```
time python -c "import diffusers"

real    0m0.417s
user    0m0.714s
sys     0m0.499s
```
Could you verify you're seeing is similar numbers after the changes?

patrickvonplaten commented:
Ok I think I found one last package that slows down the import which is `xformers`: https://github.com/facebookresearch/xformers . I'd say most of our users have `xformers` installed - can you try to also make sure `xformers` is not important by default?

DN6 commented:
@patrickvonplaten changed it so that we use `_torch_version = importlib_metadata.version("torch")` instead of `import torch; torch.__version__` when checking for xformers availability.

patrickvonplaten commented:
@DN6 could you take a look at the final failing tests or do you need help here?

DN6 commented:
@patrickvonplaten 
Fixed conflicts and added Wuerstchen.

Seeing some strange behaviour in the `Fast tests for PRs / Fast PyTorch Models & Schedulers CPU tests`. 

When running 

```shell
python -m pytest -n 2 --max-worker-restart=0 --dist=loadfile     -s -v -k "not Flax and not Onnx"  tests/models tests/schedulers
```
All tests pass

The failures occur when `tests/others` is also included. I isolated it to this test `tests/others/test_check_copies.py`

If you run this test with the model/scheduler test, you get a few failures in the scheduler tests. 

```
FAILED tests/schedulers/test_schedulers.py::SchedulerBaseTests::test_save_load_compatible_schedulers - AssertionError: assert 'The config a...ation file.\n' == 'The config a...ation file.\n'
FAILED tests/schedulers/test_schedulers.py::SchedulerBaseTests::test_save_load_from_different_config - AssertionError: assert 'The config a...ation file.\n' == ''
FAILED tests/schedulers/test_schedulers.py::SchedulerBaseTests::test_save_load_from_different_config_comp_schedulers - assert 'The config a...ult values.\n' == "{'f'} was no...ult values.\n"
```  
Any thoughts on what's happening? I can't seem to figure it out.

DN6 commented on utils/check_copies.py:
```diff
@@ -29,15 +28,6 @@
 REPO_PATH = "."
 
 
-# This is to make sure the diffusers module imported is the one in the repo.
-spec = importlib.util.spec_from_file_location(
-    "diffusers",
-    os.path.join(DIFFUSERS_PATH, "__init__.py"),
-    submodule_search_locations=[DIFFUSERS_PATH],
-)
-diffusers_module = spec.loader.load_module()
-
-
```
@patrickvonplaten This section is causing the failing fast tests for models and schedulers. `diffusers_module` doesn't appear to be used anywhere. So I removed the snippet. 

LMK if there's something I've missed here.

patrickvonplaten replied:
That works!

patrickvonplaten commented:
Alright, great job - let's get this one in to avoid any more merge conflicts.
Over the next days, let's be aware of test failures that might be related to this cc @DN6

Abhinay1997 commented:
@DN6, Looks like `StableDiffusionGLIGENPipeline` is added twice (See Line 240) in `src/diffusers/__init__.py`. I don't think it'll affect anything but just letting you know.
